File: RangeCalcCon/internal/rangemath/rangemath.go
// Copyright (c) 2026 John Dovey <dovey.john@gmail.com>
//
// Package rangemath implements simple and full triangulation range estimates.
package rangemath

import (
	"fmt"
	"math"
)

// Method selects the calculation approach.
type Method int

const (
	// MethodSimple uses range = baseline * cot(theta) where theta is the
	// circular bearing difference. Approximate; assumes favourable geometry.
	MethodSimple Method = iota
	// MethodFull uses the law of sines with a known baseline bearing.
	MethodFull
)

// Result holds ranges from each reference and the angle at the target.
type Result struct {
	RangeFromRef1 float64
	RangeFromRef2 float64
	AngleAtTarget float64
	Method        Method
}

// CircularAngleDiff returns the smallest angle between two compass bearings
// in the open-closed interval (0, 180].
func CircularAngleDiff(a, b float64) float64 {
	diff := math.Abs(a - b)
	if diff > 180 {
		diff = 360 - diff
	}
	return diff
}

func degToRad(deg float64) float64 {
	return deg * math.Pi / 180
}

// Simple computes an approximate range from two bearings and baseline length.
// Valid when 0 < circular(|b1-b2|) < 90.
func Simple(b1, b2, baseline float64) (Result, error) {
	if baseline <= 0 {
		return Result{}, fmt.Errorf("baseline distance must be greater than zero")
	}
	theta := CircularAngleDiff(b1, b2)
	if theta == 0 {
		return Result{}, fmt.Errorf("bearings are identical — cannot calculate range")
	}
	if theta >= 90 {
		return Result{}, fmt.Errorf("angle difference too large for simple mode (need < 90°); try full triangulation")
	}

	// cot(theta) = tan(90 - theta)
	rangeVal := baseline * math.Tan(degToRad(90-theta))
	return Result{
		RangeFromRef1: rangeVal,
		RangeFromRef2: rangeVal,
		AngleAtTarget: theta,
		Method:        MethodSimple,
	}, nil
}

// Full computes ranges via the law of sines using baseline bearing (Ref1→Ref2).
func Full(b1, b2, baseline, baselineBearing float64) (Result, error) {
	if baseline <= 0 {
		return Result{}, fmt.Errorf("baseline distance must be greater than zero")
	}

	angleAtTarget := CircularAngleDiff(b1, b2)
	if angleAtTarget == 0 {
		return Result{}, fmt.Errorf("bearings are identical — cannot calculate range")
	}

	// Interior angle at Ref1: between baseline (Ref1→Ref2) and LOB (Ref1→Target)
	angleAtRef1 := CircularAngleDiff(baselineBearing, b1)
	angleAtRef2 := 180 - angleAtTarget - angleAtRef1

	if angleAtRef1 <= 0 || angleAtRef2 <= 0 {
		return Result{}, fmt.Errorf("invalid geometry — target not possible with these bearings")
	}

	sinTarget := math.Sin(degToRad(angleAtTarget))
	if math.Abs(sinTarget) < 1e-12 {
		return Result{}, fmt.Errorf("degenerate triangle (angle at target near 0° or 180°)")
	}

	r1 := baseline * math.Sin(degToRad(angleAtRef2)) / sinTarget
	r2 := baseline * math.Sin(degToRad(angleAtRef1)) / sinTarget
	if r1 <= 0 || r2 <= 0 {
		return Result{}, fmt.Errorf("invalid geometry — computed range is not positive")
	}

	return Result{
		RangeFromRef1: r1,
		RangeFromRef2: r2,
		AngleAtTarget: angleAtTarget,
		Method:        MethodFull,
	}, nil
}

// MethodName returns a short label for the method.
func MethodName(m Method) string {
	switch m {
	case MethodFull:
		return "Full triangulation"
	default:
		return "Simple"
	}
}

File: RangeCalcAndroid/app/src/main/java/com/jdovey/rangecalc/rangemath/RangeMath.kt
package com.jdovey.rangecalc.rangemath

import kotlin.math.abs
import kotlin.math.sin
import kotlin.math.tan
import kotlin.math.PI

/** Selects the calculation approach. */
enum class Method {
    /** range = baseline * cot(theta) where theta is the circular bearing difference. */
    SIMPLE,

    /** Law of sines with a known baseline bearing. */
    FULL,
}

/** Ranges from each reference and the angle at the target. */
data class RangeResult(
    val rangeFromRef1: Double,
    val rangeFromRef2: Double,
    val angleAtTarget: Double,
    val method: Method,
)

/** Simple and full triangulation range estimates. Ported from RangeCalcCon's rangemath package. */
object RangeMath {

    /** Smallest angle between two compass bearings, in the open-closed interval (0, 180]. */
    fun circularAngleDiff(a: Double, b: Double): Double {
        var diff = abs(a - b)
        if (diff > 180) diff = 360 - diff
        return diff
    }

    private fun degToRad(deg: Double): Double = deg * PI / 180

    /** Approximate range from two bearings and baseline length. Valid when 0 < theta < 90. */
    fun simple(b1: Double, b2: Double, baseline: Double): Result<RangeResult> {
        if (baseline <= 0) {
            return Result.failure(IllegalArgumentException("baseline distance must be greater than zero"))
        }
        val theta = circularAngleDiff(b1, b2)
        if (theta == 0.0) {
            return Result.failure(IllegalArgumentException("bearings are identical — cannot calculate range"))
        }
        if (theta >= 90) {
            return Result.failure(
                IllegalArgumentException("angle difference too large for simple mode (need < 90°); try full triangulation")
            )
        }

        // cot(theta) = tan(90 - theta)
        val rangeVal = baseline * tan(degToRad(90 - theta))
        return Result.success(
            RangeResult(
                rangeFromRef1 = rangeVal,
                rangeFromRef2 = rangeVal,
                angleAtTarget = theta,
                method = Method.SIMPLE,
            )
        )
    }

    /** Ranges via the law of sines using baseline bearing (Ref1→Ref2). */
    fun full(b1: Double, b2: Double, baseline: Double, baselineBearing: Double): Result<RangeResult> {
        if (baseline <= 0) {
            return Result.failure(IllegalArgumentException("baseline distance must be greater than zero"))
        }

        val angleAtTarget = circularAngleDiff(b1, b2)
        if (angleAtTarget == 0.0) {
            return Result.failure(IllegalArgumentException("bearings are identical — cannot calculate range"))
        }

        // Interior angle at Ref1: between baseline (Ref1→Ref2) and LOB (Ref1→Target)
        val angleAtRef1 = circularAngleDiff(baselineBearing, b1)
        val angleAtRef2 = 180 - angleAtTarget - angleAtRef1

        if (angleAtRef1 <= 0 || angleAtRef2 <= 0) {
            return Result.failure(IllegalArgumentException("invalid geometry — target not possible with these bearings"))
        }

        val sinTarget = sin(degToRad(angleAtTarget))
        if (abs(sinTarget) < 1e-12) {
            return Result.failure(IllegalArgumentException("degenerate triangle (angle at target near 0° or 180°)"))
        }

        val r1 = baseline * sin(degToRad(angleAtRef2)) / sinTarget
        val r2 = baseline * sin(degToRad(angleAtRef1)) / sinTarget
        if (r1 <= 0 || r2 <= 0) {
            return Result.failure(IllegalArgumentException("invalid geometry — computed range is not positive"))
        }

        return Result.success(
            RangeResult(
                rangeFromRef1 = r1,
                rangeFromRef2 = r2,
                angleAtTarget = angleAtTarget,
                method = Method.FULL,
            )
        )
    }

    /** Short label for the method. */
    fun methodName(m: Method): String = when (m) {
        Method.FULL -> "Full triangulation"
        Method.SIMPLE -> "Simple"
    }
}
